File: examples/ios_hello/HelloWorldApp/Hello.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

//
//  Hello.m
//  HelloWorldApp
//

#import "Hello.h"
#import <Foundation/Foundation.h>

@implementation Hello

- (Hello*)return_hello {
  return [Hello new];
}

- (NSString*)null_dereference_bug {
  Hello* hello = nil;
  return hello->_s;
}

- (NSString*)ivar_not_nullable_bug:(Hello*)hello {
  Hello* ret_hello = [hello->_hello return_hello];
  return ret_hello->_s;
}

- (NSString*)parameter_not_null_checked_bug:(Hello*)hello {
  Hello* ret_hello = [hello return_hello];
  return ret_hello->_s;
}

@end

File: infer/tests/build_systems/codetoanalyze/make/Hello.java
/*
 * Copyright (c) 2015 - present Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

package hello;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.Random;

class Hello {

  void doesNotCauseNPE() {
    Pointers.A a = Pointers.mayReturnNull(10);
    a.method();
  }

  void mayCauseNPE() {
    Random rng = new Random();
    Pointers.A a = Pointers.mayReturnNull(rng.nextInt());
    // NPE
    a.method();
  }

  void mayLeakResource() throws IOException {
    OutputStream stream = Resources.allocateResource();
    if (stream == null) {
      return;
    }

    try {
      stream.write(12);
    } finally {
      // Resource leak
    }
  }

  void twoResources() throws IOException {
    FileInputStream fis = null;
    FileOutputStream fos = null;
    try {
      fis = new FileInputStream(new File("whatever.txt"));
      fos = new FileOutputStream(new File("everwhat.txt"));
      fos.write(fis.read());
    } finally {
      if (fis != null) { fis.close(); } // Resource leak
      if (fos != null) { fos.close(); }
    }
  }

}
